lib/puppet/resource/catalog.rb

```diff
@@ -538,7 +538,10 @@ def filter(&block)
 
   # Store the classes in the classfile.
   def write_class_file
-    ::File.open(Puppet[:classfile], "w") do |f|
+    # classfile paths may contain UTF-8
+    # https://docs.puppet.com/puppet/latest/reference/configuration.html#classfile
+    classfile = Puppet.settings.setting(:classfile)
+    Puppet::FileSystem.open(classfile.value, classfile.mode.to_i(8), "w:UTF-8") do |f|
```

Comment:
That's a great question.

This implementation will do exactly what `::File.open` does - see https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_system/file_impl.rb#L30-L32

That said, it's likely not going to be what we want on Windows, as this will set Windows file attributes (read-only, archive, etc bits) rather than security permissions.  I will change this to only apply mode on non-Windows, with a note referencing the ticket that fixes our `Puppet::FileSystem` implementation to use the mode value for permissions (I think we have a ticket somewhere for that)